### DISCUSSION TITLE
QUESTION: How to define the Generic Access Profile (during BlessServer creation)?

### DISCUSSION THREAD

**Discussion by WouterJD** [kevincar/bless#76] [General]
* bless version: current
* Python version: 3
* Operating System: Windows 10, Raspberry

### Description

`BlessServer(name=myServiceName, loop=self.loop)` creates the BlessServer;
application services are added with `self.BlessServer.add_gatt(myGattDefinition)`

### What is already working

The service is created and can be inspected (using bleak) and the Generic Access Profile is present (it's not in the myGattDefinition) so it appears to be created implicitly:
```
Service: 00001800-... (Handle: 1): Generic Access Profile
        Characteristic: handle=   2, uuid=00002a00-0000-1000-8000-00805f9b34fb props=['read'], value="rpiServer"
        Characteristic: handle=   4, uuid=00002a01-0000-1000-8000-00805f9b34fb props=['read'], value="00 00"
```

rpiServer is the HOSTNAME of the computer, not the name provided to the BlessServer-creation,

### What is NOT yet working --- QUESTION:
How to set the two characteristics:
```
DeviceNameUUID = '00002a00-0000-1000-8000-00805f9b34fb'
AppearanceUUID = '00002a01-0000-1000-8000-00805f9b34fb'
```
?

**Comment by WouterJD** [comment]
Hi @kevincar could you have a look into this please?
It would be very helpful for me to understand how to set the two General Access Profile characteristics, so that my FTMS is publsihed witha correct name instead of the hostname where the FTMS is running.

Thanks for you support

**Comment by kevincar** [comment]
Since this doesn't appear to be an issue just yet, let's move this to a discussion

**Comment by WouterJD** [comment]
Hi @kevincar thanks for response.

I cannot judge whether this is an issue or not
Creating a BlessServer with `BlessServer(name=myServiceName, loop=self.loop)` creates the BlessServer, but the servicename is not the provided myServiceName but the hostname where the service is running.

I'd be glad to know what to do that myServiceName is seen as the FTMS DeviceName.

**Comment by kevincar** [comment]
For your main question for how to set those characteristics using a generic attribute profile tree, you can do something like the following:

```python
    # Instantiate the server
    gatt: Dict = {
        "<YOUR SERVICE UUID>": {
            "00002a00-0000-1000-8000-00805f9b34fb": {  # <-- Characteristic IDs are the Key names
                "Properties": (GATTCharacteristicProperties.read |
                               GATTCharacteristicProperties.indicate),
                "Permissions": (GATTAttributePermissions.readable),
                "Value": "My Device Name"
            },
            "00002a01-0000-1000-8000-00805f9b34fb": {
                "Properties": (GATTCharacteristicProperties.read |
                               GATTCharacteristicProperties.indicate),
                "Permissions": (GATTAttributePermissions.readable),
                "Value": "My Appearance UUID"
            }
        }
    }
    my_service_name = "Test Service"
    server = BlessServer(name=my_service_name, loop=loop)
    await server.add_gatt(gatt)
    await server.start()
```

**Reply by WouterJD** [reply]
> ```
> "<YOUR SERVICE UUID>"
> ```
So I should not use the Generic Access Profile UUID here?

**Reply by kevincar** [reply]
You should specify your service UUID. Bluetooth Low-Energy uses the Generic Access Profile to structure the data that it broadcasts in the form of services and characteristics. Each service and each characteristic is identified by a unique identifier (UUID). There are predefined characteristic UUID, but not (to my knowledge) service UUIDs. You provide a unique value for your application.

**Comment by WouterJD** [comment]
Thanks, will try asap.
I got errors when trying this - I will retry.
👍

**Comment by WouterJD** [comment]
Hi @kevincar

I have tried a series of experiments, as explained below.

Behaviour on Windows and Raspberry is different:
- uuid = GenericAccessUUID is not allowed at either side; add_gatt() returns an error
- uuid = \<YOUR SERVICE UUID\> is allowed, but not recognized as such by clients
  - Raspberry server causes duplicated UUID at client-side and then FTMS cannot be retrieved anymore (confuses bleak)
  - Windows server accepts but the client does not understand it as a correct DeviceName (client = Trainer Road)

I have the impression that the underlaying libraries need a call to create the General Access Profile, but I have no knowledge at that level.
Your view on this is highly appreciated.


```
    # On Windows 10:
    #     bc.sGenericAccessUUID: { ... containing the two characteristics ... }
    #       results in:
    #           clsBleServer._Server(); add_gatt() exception 'NoneType' object has no attribute 'add_advertisement_status_changed'
    #
    #     bc.sFitnessMachineUUID: { ... containing the two characteristics ... }
    #       results in that characteristic can be read by client, but Trainer Road does still provide TP-T430 as service name.
    #
    #     bc.sGenericAccessUUID_private: { ... containing the two characteristics ... }
    #       has the same effect.
    #       bleBleak considers the service to be "vendor specific"
    #
    # On Raspberry:
    #     bc.sGenericAccessUUID: { ... containing the two characteristics ... }
    #       results in:
    #            clsBleServer._Server(); start() exception org.bluez.Error.Failed: Failed to create entry in database
    #
    #     bc.sGenericAccessUUID_private: { ... containing the two characteristics ... }
    #       This confuses bleak at the Windows 10 side (duplicate keys, FTMS not found)

    "<YOUR SERVICE UUID>": {            # This line is modified as explained above
        bc.cDeviceNameUUID: {
            "Properties":   (GATTCharacteristicProperties.read),
            "Permissions":  (GATTAttributePermissions.readable | GATTAttributePermissions.writeable),
            "Value":        b'FortiusAntTrainer',
            "value":        b'FortiusAntTrainer',
            "Description":  bc.cDeviceNameName
        },
        bc.cAppearanceUUID: {
            "Properties":   (GATTCharacteristicProperties.read),
            "Permissions":  (GATTAttributePermissions.readable | GATTAttributePermissions.writeable),
            "Value":        b'\x80\x00',
            "value":        b'\x80\x00',
            "Description":  bc.cAppearanceName
        },
    },

```

**Reply by kevincar** [reply]
Focusing on the windows backend first: 

1. What is `clsBleServer._Server()` doing?
2. What argument are you passing to `add_gatt()`? Is it this:

```
"<YOUR SERVICE UUID>": {
        bc.cDeviceNameUUID: {
            "Properties":   (GATTCharacteristicProperties.read),
            "Permissions":  (GATTAttributePermissions.readable | GATTAttributePermissions.writeable),
            "Value":        b'FortiusAntTrainer',
            "value":        b'FortiusAntTrainer',
            "Description":  bc.cDeviceNameName
        },
        bc.cAppearanceUUID: {
            "Properties":   (GATTCharacteristicProperties.read),
            "Permissions":  (GATTAttributePermissions.readable | GATTAttributePermissions.writeable),
            "Value":        b'\x80\x00',
            "value":        b'\x80\x00',
            "Description":  bc.cAppearanceName
        },
    }
```

3. What is the service ID? This error `exception 'NoneType' object has no attribute 'add_advertisement_status_changed'` is likely because the service is failing to be created.
4. Why do you have both `value` and `Value`? the upper case `Value` will be ignored when passed to the `add_gatt()` method.

**Comment by WouterJD** [comment]
Hi @kevin, thanks for being with me.
I have published the code as a pull-request, which should be executable on your environment. I have tested on Windows/Raspberry running Server/Client or Client/Server.

1. Has the same function as run() in gattserver.py; not waiting for a trigger but having a timed-loop, so that actions can be taken while waiting.

2. add_gatt() receives the structure as supplied by FTMSserver.py; FitnessMachineGatt:
https://github.com/WouterJD/bless/blob/FTMS-example/examples/FTMSserver.py#L161
The page above it are all commented for the duration of the test. Candidate for removal when final version is clear.

3. The ServiceID = bc.sFitnessMachineUUID = "00001826-0000-1000-8000-00805f9b34fb"
https://github.com/WouterJD/bless/blob/FTMS-example/examples/FTMSconstants.py#L50
The error occurs when ServiceID = '00001800-0000-1000-8000-00805f9b34fb' (Generic Access Profile) **is added** to the structure, which appears not to be allowed, since that Service ID is created by default.

4. value and Value; because of the confusion which of the two should be the case. See https://github.com/kevincar/bless/issues/67. I use both so .value and .Value work. To be cleaned for final version.

I hope this helps in understanding my issue.

**Comment by WouterJD** [comment]
Hi @kevincar https://github.com/kevincar/bless/pull/77 contains the code to reproduce the issue; I hope you find time to give it a glance.
Most appreciated!

**Reply by kevincar** [reply]
Your code actually worked for me *as is* on both Raspberry Pi and Windows. 

This is a screenshot from interrogating the services using an app on my mobile device
<img src="https://user-images.githubusercontent.com/3383537/159706880-6202a266-8df2-4215-ab8a-ed112e02e30f.jpg" width="250px" />

Do you know what version of bless you are using when you run your example code? Perhaps reinitializing your python environment and ensuring you test by installing bless from the `develop` branch may help.

**Reply by WouterJD** [reply]
> Your code actually worked for me _as is_ on both Raspberry Pi and Windows.

That's great and a great app, so we now talk about the same and a client you trust to display correct values.

The DeviceName is the recognizable name of the device and can usually be customized by the owner.
Currently the device name contains the hostname where bless is running.
I would expect it to contain myServiceName as in `self.BlessServer = BlessServer(name=self.myServiceName, loop=self.loop)`


### FTMSserver.py running on raspberry

QUESTION:
1. DeviceName = **rpiServer** but not the provided name in `self.BlessServer = BlessServer(name=self.myServiceName, loop=self.loop)`
2. Appearance = 0x0000

<img src="https://user-images.githubusercontent.com/47486941/159791533-ef3796f1-b647-4ef9-aa2c-5656e046b781.png" width="250px" /> <img src="https://user-images.githubusercontent.com/47486941/159792271-5326f283-a5d8-4b7c-9d25-592f9f28bb51.png" width="250px" />

---
---

### FTMSserver.py running on Windows 10

QUESTION:
1. DeviceName = **LT-T430** but not the provided name in `self.BlessServer = BlessServer(name=self.myServiceName, loop=self.loop)`
2. Appearance = 0x8000; this is different from raspberry

<img src="https://user-images.githubusercontent.com/47486941/159793118-fff3fa4f-e961-42f1-9d88-4f1afd7b9e54.png" width="250px" />

<img src="https://user-images.githubusercontent.com/47486941/159793142-90d1b05a-2c2e-45d9-bcfd-4d16f0655b2f.png" width="250px" />

**Comment by WouterJD** [comment]


**Reply by kevincar** [reply]
I use [this app](https://punchthrough.com/lightblue/) to test `bless`

**Comment by WouterJD** [comment]
I use version= "0.2.3"

I also downloaded `develop` branch, but have the impression that the version is the same, is that correct?
Is there a way to print the current version of bless?

(As in bleak: `from bleak import __version__ as bleakVersion`)

**Reply by kevincar** [reply]
We could certainly implement that. One way to check which version you're using is `pip freeze | grep bless`

The develop branch has many issues fixed that haven't yet been pushed to the master branch

**Comment by WouterJD** [comment]
Below two FTMSserver.py instances, one running on Windows (LT-T430) and the other on raspberry.
Do you know what is the reason that the raspberry does not display a name?

I have upgraded the raspberry with command `pip install --force-reinstall git+https://github.com/kevincar/bless.git@develop` which makes no difference.

![image](https://user-images.githubusercontent.com/47486941/159795296-65702f37-c671-4b31-aa9a-018f75c62881.png)

**Reply by kevincar** [reply]
I think this is likely due to the linux backend of the package is still growing in development (#60), this may be an ongoing issue. But the data transfer between characteristics should still be functional.

**Comment by WouterJD** [comment]
I'm getting closer, but not there yet.
On computers, the DeviceID is the hostname. For trainers this is not an issue, since there is one service and setting the hostname to the desired trainers name is quite straight forward.

When the DeviceId should be different from the hostname, [this article](https://stackoverflow.com/questions/26299053/changing-raspberry-pi-bluetooth-device-name ) helps. But still, on raspberry the ADDRESS is shown in the list of devices and not the PRETTY_HOSTNAME.

The ennoying thiung is that, when using NodeJS, the correct name is displayed AND it even remains present when switching to bless. So it seems some system- or (session-) level setting.

Will keep you posted.

> But the data transfer between characteristics should still be functional.

Could you please suggest **HOW to change the Generic Access Profile characteristics through bless**, I have tried multiple ways but no success yet.

**Reply by kevincar** [reply]
@WouterJD I've attempted to address this in the `60-dbus-next` branch. I tested it out and the service name should now be broadcasting appropriately on Raspberry PI.

**Comment by WouterJD** [comment]
My results for today:

[This article](https://stackoverflow.com/questions/26299053/changing-raspberry-pi-bluetooth-device-name) provides usefull information to change the DeviceID on raspberry and that works. 

**But** there is also something like the "Local Name" which is not the same as the DeviceID, see below screenshots. "Unnamed" is not very elegant and it seems something is left uninitialized.

On raspberry, there is no local device name although the DeviceID is available:
![image](https://user-images.githubusercontent.com/47486941/159939451-8ce2111b-5d3d-406d-9988-0dc0644917d3.png)

On windows, the local device name is also filled with the hostname (LT-T430):
![image](https://user-images.githubusercontent.com/47486941/159939678-2ae57760-85ad-4fed-9533-ad56352c377f.png)

**Comment by WouterJD** [comment]
Good news, I will test Monday evening👍👍

**Comment by WouterJD** [comment]
Test block, see issue #60 [comment](https://github.com/kevincar/bless/issues/60#issuecomment-1081012340) to be sorted out first

**Comment by WouterJD** [comment]
Sorry for taking so long. With modified access rights, I can test.
The FortiusAntTrainer is still seen as a bluetooth-ADDRESS, not the name, like before.

If this is what is possible, please let me know - I can live with it and the users of FortiusAnt also can.

![image](https://user-images.githubusercontent.com/47486941/159939451-8ce2111b-5d3d-406d-9988-0dc0644917d3.png)
